src/bgen/TypeManager.cs

```diff
@@ -413,4 +433,124 @@ public TypeManager (BindingTouch bindingTouch, Assembly apiAssembly, Assembly co
 			NSDirectionalEdgeInsets = ConditionalLookup (platformAssembly, "UIKit", "NSDirectionalEdgeInsets");
 		}
 	}
+
+	public string PrimitiveType (Type t, bool formatted = false)
+	{
+		if (t == System_Void)
+			return "void";
+		if (t == System_Int32)
+			return "int";
+		if (t == System_Int16)
+			return "short";
+		if (t == System_Byte)
+			return "byte";
+		if (t == System_Float)
+			return "float";
+		if (t == System_Boolean)
+			return "bool";
+		if (t == System_Char)
+			return "char";
+		if (t == System_nfloat)
+			return "nfloat";
+
+		return formatted ? FormatType (null, t) : t.Name;
+	}
+
+	public string FormatType (Type? usedIn, Type type)
+	{
+		return FormatTypeUsedIn (usedIn?.Namespace, type);
+	}
+
+	public string FormatType (Type? usedIn, Type type, bool protocolized)
+	{
+		return FormatTypeUsedIn (usedIn?.Namespace, type, protocolized);
+	}
+
+	public string FormatType (Type? usedIn, string @namespace, string name)
+	{
+		string tname;
+		if ((usedIn is not null && @namespace == usedIn.Namespace) || BindingTouch.NamespaceManager.StandardNamespaces.Contains (@namespace))
+			tname = name;
+		else
+			tname = "global::" + @namespace + "." + name;
+
+		return tname;
+	}
+
+	public string FormatTypeUsedIn (string? usedInNamespace, Type? type, bool protocolized = false)
+	{
+		if (type is null)
```

Comment:
I added a condition for null on account of type.GetElementType() (called in 524) having the possibility of null values (example in the mono docs even have a null example, interesting)